Allow methods to be named after keywords

vexx32 opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Adds a check for whether a token has the `TokenFlags.Keyword` flag or is `TokenKind.DynamicKeyword` if it isn't tokenized as `TokenKind.Identifier` when examining possible tokens to use for the method name. This means that any keyword may be used as a method name, with the only remaining exceptions being  `hidden` or `static`. 

As these keywords are used in method declarations, I feel these remaining unusable as a method name is fairly sensible. 

Additionally, once a usable method token is found, additionally we remove the token flags to ensure we don't get the keyword highlighting on such method names.

## PR Context

Fixes #9106
Fixes #2217

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

vexx32 committed:
:tada: Allow keywords as method names.
Allowing all keywords to be used as method names.
This shouldn't clash with hidden/static as these are required to be
provided first, before type declarations and method names.

vexx32 commented:
@daxian-dbw @iSazonov Not sure where to put tests for this. Any suggestions? 🙂

iSazonov commented:
Perhaps Parser.Tests.ps1 and Parsing.Tests.ps1

vexx32 commented:
Added a test in Parser.Tests.ps1

I looked at Parsing.Tests.ps1 but I'm frankly very confused as to how that testing goes where classes are concerned, as there are none in there currently. I managed to piece together what it's doing where functions are concerned, but I've yet to figure out how it's supposed to test things where classes are concerned.

vexx32 commented:
Windows agent failed on.... something very odd. Might need to be kicked again. 😄

PoshChan commented:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Permits class methods to be named after keywords
```powershell
Expected no exception to be thrown, but an exception "Exception calling "Invoke" with "0" argument(s): "At line:8 char:18
+                 [hidden] $Data = "Secrets"
+                  ~~~~~~
Unable to find type [hidden]."" was thrown from /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1:80 char:17
    +                 $PowerShell.AddScript($command).Invoke()
    +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1031
1031:             { ExecuteCommand $ClassDefinition } | Should -Not -Throw
```

PoshChan commented:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
ParserTests (admin\monad\tests\monad\src\engine\core\ParserTests.cs).Permits class methods to be named after keywords
```powershell
Expected no exception to be thrown, but an exception "Exception calling "Invoke" with "0" argument(s): "At line:8 char:18
+                 [hidden] $Data = "Secrets"
+                  ~~~~~~
Unable to find type [hidden]."" was thrown from /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1:80 char:17
    +                 $PowerShell.AddScript($command).Invoke()
    +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
at <ScriptBlock>, /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Parser/Parser.Tests.ps1: line 1031
1031:             { ExecuteCommand $ClassDefinition } | Should -Not -Throw
```

vexx32 commented:
🤔 OK, something seems to be up with the Windows build agent. @TravisEz13 have you seen this happen before? 
```
##[section]Starting: Download build artifacts
==============================================================================
Task         : Download build artifacts
Description  : Download files that were saved as artifacts of a completed build
Version      : 0.153.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/download-build-artifacts
==============================================================================
Downloading artifacts for build: 25058
Linked artifacts count:  3
Downloading artifact build from: https://powershell.visualstudio.com//2972bb5c-f20c-4a60-8bd9-00ffe9987edc/_apis/build/builds/25058/artifacts?artifactName=build&$format=zip
Downloading artifact testResults from: https://powershell.visualstudio.com//2972bb5c-f20c-4a60-8bd9-00ffe9987edc/_apis/build/builds/25058/artifacts?artifactName=testResults&$format=zip
Downloading artifact xunit from: https://powershell.visualstudio.com//2972bb5c-f20c-4a60-8bd9-00ffe9987edc/_apis/build/builds/25058/artifacts?artifactName=xunit&$format=zip
Skipped processing item 
Skipped processing item 
Skipped processing item 
Successfully downloaded artifacts to D:\a\1\a
Total Files: 1, Processed: 0, Skipped: 1, Failed: 0, Download time: 0.015 secs, Download size: 0Bytes

D:\a\_tasks\DownloadBuildArtifacts_a433f589-fce1-4460-9ee6-44a624aeb1fb\0.153.1\node_modules\artifact-engine\Engine\artifactEngine.js:157
    throw err;
    ^
Failed to extract package with error Error: ENOENT: no such file or directory, open 'D:\a\1\a\build.zip'
Total Files: 1, Processed: 0, Skipped: 1, Failed: 0, Download time: 0.015 secs, Download size: 0Bytes
Total Files: 1, Processed: 0, Skipped: 1, Failed: 0, Download time: 0.015 secs, Download size: 0Bytes
##[error]artifact-engine: unhandled rejection Failed to extract package with error Error: ENOENT: no such file or directory, open 'D:\a\1\a\build.zip'
##[error]Unhandled: undefined
##[error]Unhandled: undefined
##[error]artifact-engine: unhandled exception Failed to extract package with error Error: ENOENT: no such file or directory, open 'D:\a\1\a\build.zip'
##[error]Exit code 7 returned from process: file name 'C:\agents\2.152.1\externals\node\bin\node.exe', arguments '"D:\a\_tasks\DownloadBuildArtifacts_a433f589-fce1-4460-9ee6-44a624aeb1fb\0.153.1\main.js"'.
##[section]Finishing: Download build artifacts
```

daxian-dbw commented:
@vexx32 The class related tests are located at `test\powershell\Language\Classes`. Feel free to move your test there :)

vexx32 committed:
:white_check_mark: add test for class name parsing

PoshChan commented:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
Method Name Test.Permits class methods to be named after keywords
```powershell
Expected no exception to be thrown, but an exception "The term 'ExecuteCommand' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again." was thrown from /home/vsts/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1:443 char:11
    +         { ExecuteCommand $ClassDefinition } | Should -Not -Throw
    +           ~~~~~~~~~~~~~~.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1: line 443
443:         { ExecuteCommand $ClassDefinition } | Should -Not -Throw
```

vexx32 commented:
Oh, of course. Let's reframe that.

PoshChan commented:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
Method Name Test.Permits class methods to be named after keywords
```powershell
Expected no exception to be thrown, but an exception "The term 'ExecuteCommand' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again." was thrown from /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1:443 char:11
    +         { ExecuteCommand $ClassDefinition } | Should -Not -Throw
    +           ~~~~~~~~~~~~~~.
at <ScriptBlock>, /Users/vsts/agent/2.152.1/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1: line 443
443:         { ExecuteCommand $ClassDefinition } | Should -Not -Throw
```

vexx32 committed:
:white_check_mark: use more direct test

vexx32 commented:
That should do it. Not sure what's up w/ the windows agent though @daxian-dbw

daxian-dbw commented:
@vexx32 We have reported the issue to Azure DevOps. Let's wait for their response.

iSazonov commented on test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1:
```diff
@@ -416,6 +416,35 @@ Describe 'Property Attributes Test' -Tags "CI" {
         It "second value should be b" { $v.ValidValues[1] | Should -Be 'b' }
 }
 
+Describe 'Method Name Test' -Tags "CI" {
```
Please make the title more informative like "Method names can be keywords".

iSazonov reviewed (approved):
LGTM with one minor comment.

vexx32 committed:
:memo: More descriptive describe name for test

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4514,8 +4514,11 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
+            if (token.Kind == TokenKind.Identifier
+                || token.TokenFlags.HasFlag(TokenFlags.Keyword))
             {
+                token.TokenFlags = TokenFlags.None;
```
It would be great if we can use [`SimpleNameRule`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Parser.cs#L2131) here to get the method name, just like how we parse the class name, so the token we got back is accurate. See the two `class` tokens below for an example.
```
$tokens = $null
$s = [System.Management.Automation.Language.Parser]::ParseInput("class class {  }", [ref]$tokens, [ref]$null)
PS:20> $tokens

Text       : class
TokenFlags : Keyword
Kind       : Class
HasError   : False
Extent     : class

Text       : class
TokenFlags : TypeName
Kind       : Identifier
HasError   : False
Extent     : class

...
```
See how `ClassDefinitionRule` does it:
https://github.com/PowerShell/PowerShell/blob/981c990d747ae97f56a6c7474e5d2ef04bf237c5/src/System.Management.Automation/engine/parser/Parser.cs#L4156-L4169

However, that would require we resync the token because the token has already been read before setting `WantSimpleName`, and `Resync` is relatively expensive.
I guess I'm fine we just check for `Keyword` and `DynamicKeyword` here as a special case of `SimpleNameRule`, but I'm not sure if `SimpleNameRule` covers some other token cases as well.

Note that, you handled `keyword` but not the `DynamicKeyword`, see the code in `Tokenizer`:
https://github.com/PowerShell/PowerShell/blob/981c990d747ae97f56a6c7474e5d2ef04bf237c5/src/System.Management.Automation/engine/parser/tokenizer.cs#L4371-L4385

vexx32 replied:
Heh, I was poking at that, seeing if there was a way to make a tokenizer exception for keywords in the specific context of a method name, but I couldn't find anything that seemed to work.

But yeah, we can definitely add `DynamicKeyword` as well. Good point!

daxian-dbw replied:
The way `SimpleNameRule` works is to set the special `WantSimpleName` flag and then go on with the next tokenizing work, which requires that no token as been peeked yet.
In our case here, we already peeked the token, which means the token was read with `WantSimpleName = false`. To use `SimpleNameRule`, we will have to resync back to the start offset of the peeked token.

vexx32 replied:
Hmm, I see.

Let me examine this a bit more closely. There's probably no better way, but I want to make sure. 🙂

daxian-dbw replied:
In case there is no better way (meaning we have to resync in order to leverage `WantSimpleName` flag in the tokenizer), then maybe it makes sense to make it a private utility method to alter a token that is to be used as a simple name, so the code can be reused in other rules.

vexx32 replied:
That was my backup thought as well! 🙂

vexx32 replied:
Yep, looks like that'll have to be the way forward, short of recoding that whole section of the tokenizer _and_ parser. Probably only to arrive at a similar issue somewhere else anyway. 😄

iSazonov replied:
As for `DynamicKeyword`  - Could this complicate the readability of scripts too much?
For keywords PSSA should caution I think.

vexx32 committed:
:sparkles: Add method for simple name tokens
If a token could be a simple name, it removes the token flags and
returns true.
Otherwise, no changes are made and it returns false.

vexx32 commented:
@daxian-dbw I think that addresses all your concerns.

Azure's build agents are still a bit under the weather though, unfortunately. 🙂

PoshChan commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4514,7 +4514,7 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
                 }
```
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
Enter-PSHostProcess tests.By CustomPipeName.Can enter, exit, and re-enter using CustomPipeName
```powershell
Expected $true, because The script was able to re-enter another process and grab the pipe of 'akl3gcul.0uh'., but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Core\Enter-PSHostProcess.Tests.ps1: line 170
170:                     Should -BeTrue -Because "The script was able to re-enter another process and grab the pipe of '$pipeName'."
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-macos`
Testing Method Names can be Keywords.Permits class methods to be named after keywords
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1: line 472
472:         $Results = Invoke-PowerShell -Script $TestScript

```
Testing Method Names can be Keywords.Permits class methods to be named after DynamicKeywords
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1: line 520
520:         Invoke-PowerShell -Script $DefineKeyword

```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-windows`
Testing Method Names can be Keywords.Permits class methods to be named after keywords
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Classes\Scripting.Classes.BasicParsing.Tests.ps1: line 472
472:         $Results = Invoke-PowerShell -Script $TestScript

```
Testing Method Names can be Keywords.Permits class methods to be named after DynamicKeywords
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, D:\a\1\s\test\powershell\Language\Classes\Scripting.Classes.BasicParsing.Tests.ps1: line 520
520:         Invoke-PowerShell -Script $DefineKeyword

```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-linux`
Testing Method Names can be Keywords.Permits class methods to be named after keywords
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1: line 472
472:         $Results = Invoke-PowerShell -Script $TestScript

```
Testing Method Names can be Keywords.Permits class methods to be named after DynamicKeywords
```powershell
You cannot call a method on a null-valued expression.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1: line 520
520:         Invoke-PowerShell -Script $DefineKeyword

```

PoshChan commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4514,7 +4514,7 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
                 }
```
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
Clean up open Runspaces when exit powershell process.PowerShell process should not freeze at exit
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Api/BasicEngine.Tests.ps1: line 53
53:         Wait-UntilTrue -sb { $process.HasExited } -TimeoutInMilliseconds 5000 -IntervalInMilliseconds 1000 | Should -BeTrue
```

vexx32 replied:
Seems like another transient failure.... very odd.

iSazonov replied:
Restarted Macos and Linux CIs.

vexx32 commented:
@daxian-dbw could you update your review when you have a moment? 💖

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4560,6 +4560,19 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
+            }
+
+            return false;
+        }
```
When a keyword is tokenized as a simple name, the token kind should be `Identifier` and the `TokenFlags` should be set appropriately (in our case here, should be `MemberName` I believe)

The following is an example where the class name is parsed using `SimpleNameRule`. Take a look at the second `class` token.
```
$tokens = $null
$s = [System.Management.Automation.Language.Parser]::ParseInput("class class {  }", [ref]$tokens, [ref]$null)
PS:20> $tokens

Text       : class
TokenFlags : Keyword
Kind       : Class
HasError   : False
Extent     : class

Text       : class
TokenFlags : TypeName
Kind       : Identifier
HasError   : False
Extent     : class
...
```
So here, the utility method probably should be `TryUseAsSimpleName(Token old, out Token new)`, where `new` should be a new token that has `Kind = Identifier` and `TokenFlags = MemberName`. But then, the new token needs to be put in the `Tokenizer.TokenList` and replace the old token (see `Tokenizer.ReplaceSavedTokens`) ... now it would be more or less the same as `Resync` in terms of the overhead. Given that, maybe it makes sense to just `Resync` and call the `SameNameRule`?

vexx32 replied:
Hmm, maybe?

When I checked the difference in token kind and flags here, I noted that there appear to be no flags set at all for method names. Is MemberName used for properties instead, perhaps? :thinking:

daxian-dbw replied:
I'm moving away from the idea to `Resync` and use `SimpleNameRule` here. The [`FunctionDeclarationRule`](https://github.com/PowerShell/PowerShell/blob/4ae04335a498ccbc8f3a2eb16bbd2dc54f0956e1/src/System.Management.Automation/engine/parser/Parser.cs#L5234-L5272) accepts the token as is for the function name.

Note that for function declaration, a wider range of tokens (_more than just simple names, e.g. number token is also valid_) are considered valid and that's different from what a class method name can take, but I guess it's OK to not have the class name token shown up as `Identifier` kind. We can go back to the `Resync` and `SimpleNameRule` if this becomes a problem.

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4560,6 +4560,19 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
             return null;
         }
 
+        private bool CanBeSimpleNameToken(ref Token token)
```
How about naming it as `TryUseTokenAsSimpleName`?

vexx32 replied:
```suggestion
        private bool TryUseTokenAsSimpleName(Token token)
```

daxian-dbw commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4560,6 +4560,19 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
             return null;
         }
 
+        private bool CanBeSimpleNameToken(ref Token token)
```
Also, `Token` is a reference type, not a value type. So `ref` here is not needed.

vexx32 commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -4514,7 +4514,7 @@ private MemberAst ClassMemberRule(string className, out List<Ast> astsOnError)
             }
 
-            if (token.Kind == TokenKind.Identifier)
+            if (CanBeSimpleNameToken(ref token))
```
```suggestion
            if (TryUseTokenAsSimpleName(token))
```

vexx32 committed:
Apply suggestions from code review

Update method name for simplenametoken

iSazonov reviewed (commented):
Formally it is LGTM.

My main concern is about enabling DynamicKeyword. I'd postpone this until we get special request from community.
But if we accept this in the PR we should have a test for this too.

vexx32 committed:
:recycle: use [powershell] instance for tests
This will make it easier working with the class definitions

vexx32 committed:
:sparkles: Test methods with DynamicKeyword names
If a method is named after a dynamickeyword we should still allow it.

vexx32 commented:
Had to refactor the tests slightly to avoid polluting the test scope with a dynamic keyword and risk breaking something later. Changed them to use a nested `[powershell]` instance.

That should work, unless I miss my understanding of how Pester manages scoping. @iSazonov that should work for a dynamickeyword test, I think? 🙂

vexx32 commented:
@PoshChan please remind me in 1 hour.

vexx32 committed:
:white_check_mark: Adjust scoping to fix test

PoshChan commented:
@vexx32, this is the reminder you requested 1 hour. ago

vexx32 commented:
@iSazonov I think that more or less addresses everything.

I'm a little unsure about DynamicKeyword myself, but I think it makes the most sense to treat it the same as other keywords. After all, it would be very confusing for users for some arbitrary method name to stop parsing after they define a dynamic keyword -- two seemingly unrelated things, which should stay unrelated.

daxian-dbw commented on test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1:
```diff
@@ -416,6 +416,114 @@ Describe 'Property Attributes Test' -Tags "CI" {
+        Invoke-PowerShell -Script $DefineKeyword
+        Invoke-PowerShell -Script $DefineClass
+
+        Invoke-PowerShell -Script $TestScript | Should -BeExactly 'TOP SECRET'
```
It's better to reset the `DynamicKeyword` after the test by `[System.Management.Automation.Language.DynamicKeyword]::Reset()`

daxian-dbw commented on test/powershell/Language/Classes/Scripting.Classes.BasicParsing.Tests.ps1:
```diff
@@ -416,6 +416,114 @@ Describe 'Property Attributes Test' -Tags "CI" {
+        Invoke-PowerShell -Script $DefineClass
+
+        Invoke-PowerShell -Script $TestScript | Should -BeExactly 'TOP SECRET'
+    }
```
It's better to reset the `DynamicKeyword` after the test by `[System.Management.Automation.Language.DynamicKeyword]::Reset()`

vexx32 replied:
Oh, awesome! Didn't see that one there, thanks! I'll add it in an AfterAll block or something. :)

daxian-dbw reviewed (commented):
@vexx32 LGTM with one minor feedback in test.

vexx32 committed:
:white_check_mark: Address Dongbo's comment
Make sure we remove the dynamic keywords after the test.

daxian-dbw commented:
@PoshChan Please remind me in 24 hours

PoshChan commented:
@daxian-dbw, this is the reminder you requested 24 hours ago
